tensorzero-core/src/endpoints/datasets/v1/types.rs

```diff
@@ -61,19 +61,105 @@ pub struct UpdateChatDatapointRequest {
+}
+
+/// A request to update the dynamic tool parameters of a datapoint.
+#[derive(Debug, Serialize, Deserialize, Default, Clone, JsonSchema, ts_rs::TS)]
+#[ts(export, optional_fields)]
+#[export_schema]
+pub struct UpdateDynamicToolParamsRequest {
+    /// A subset of static tools configured for the function that the inference is explicitly allowed to use.
+    /// If omitted, it will be left unchanged. If specified as `null`, it will be cleared (we allow function-configured tools plus additional tools
+    /// provided at inference time). If specified as a value, it will be set to the provided value.
+    #[serde(default, deserialize_with = "deserialize_double_option")]
+    #[schemars(extend("x-double-option" = true))]
+    pub allowed_tools: Option<Option<Vec<String>>>,
+
+    /// Tools that the user provided at inference time (not in function config), in addition to the function-configured tools, that are also allowed.
+    /// Modifying `additional_tools` DOES NOT automatically modify `allowed_tools`; `allowed_tools` must be explicitly updated to include
+    /// new tools or exclude removed tools.
+    /// If omitted, it will be left unchanged. If specified as a value, it will be set to the provided value.
+    pub additional_tools: Option<Vec<ClientSideFunctionTool>>,
+
+    /// User-specified tool choice strategy.
+    /// If omitted, it will be left unchanged. If specified as `null`, we will clear the dynamic tool choice and use function-configured tool choice.
+    #[serde(default, deserialize_with = "deserialize_double_option")]
+    #[schemars(extend("x-double-option" = true))]
+    pub tool_choice: Option<Option<ToolChoice>>,
+
+    /// Whether to use parallel tool calls in the inference.
+    /// If omitted, it will be left unchanged. If specified as `null`, it will be set to `null`. If specified as a value, it will be set to the provided value.
+    #[serde(default, deserialize_with = "deserialize_double_option")]
+    #[schemars(extend("x-double-option" = true))]
+    pub parallel_tool_calls: Option<Option<bool>>,
+
+    /// Provider-specific tool configurations
+    /// If omitted, it will be left unchanged. If specified as a value, it will be set to the provided value.
+    pub provider_tools: Option<Vec<ProviderTool>>,
+}
+
+impl UpdateDynamicToolParamsRequest {
+    /// Updates the dynamic tool parameters with the provided request.
+    /// Returns `None` if existing tool params should not be updated; returns Some(DynamicToolParams) with the updated values.
+    pub fn apply_dynamic_tool_params_update(
+        self,
+        existing_tool_params: Option<DynamicToolParams>,
+    ) -> Option<DynamicToolParams> {
+        if self.allowed_tools.is_none()
+            && self.additional_tools.is_none()
+            && self.tool_choice.is_none()
+            && self.parallel_tool_calls.is_none()
+            && self.provider_tools.is_none()
+        {
+            return None;
+        }
```

Comment:
are we sure this is correct? what if the `existing_tool_params` is Some() and then all the updates are None?

Reply:
Restructured the entire function.